android-components/components/concept/awesomebar/src/main/java/mozilla/components/concept/awesomebar/AwesomeBar.kt

```diff
@@ -105,6 +105,8 @@ interface AwesomeBar {
         val onSuggestionClicked: (() -> Unit)? = null,
         val onChipClicked: ((Chip) -> Unit)? = null,
         val score: Int = 0,
+        val onSuggestionShownForEngagement: ((Position) -> Unit)? = null,
+        val onSuggestionHiddenForEngagement: (() -> Unit)? = null,
```

Comment:
I think it would make more sense to remove these and replace this workflow with a (optional, to avoid breaking changes) callback to `AwesomeBar` Composable. Something like:

```kotlin
@Composable
fun AwesomeBar(
...
    onAutoComplete: (AwesomeBar.SuggestionProviderGroup, AwesomeBar.Suggestion) -> Unit,
    onVisibleSuggestionsUpdated: (List<Pair<AwesomeBar.Suggestion, AwesomeBar.Suggestion.Position>>) -> Unit = {},
    onScroll: () -> Unit = {},
)
```

Then, in combination with my below comment, the `Suggestions` Composable would call that directly. Then, in `AwesomeBarWrapper` we would define the lambda:

```kotlin
onVisibleSuggestionsUpdated = { 
    context.components.appStore.dispatch(AppAction.UpdatePendingFxSuggestImpressions(it))
},
```

There's a bit of detail left out here, so please let me know if anything isn't clear!

Reply:
> a reporting URL, advertiser name, block ID, and category name

It looked like this was already implemented as the `Suggestion.toImpression()` extension function. The `MetricsMiddleware` should be able to just map over the visible suggestions and call that, right? Or am I missing something still?

Reply:
Ah, `Suggestion.toImpression` is an extension function on `mozilla.appservices.suggest.Suggestion`, but the provider returns `AwesomeBar.Suggestion`s.